### PR TITLE
Double buffering support

### PR DESCRIPTION
**Pull request by naoyam**
Adds `TensorView::doubleBuffer()`. See the new tests how it is used.

For an overview of the lowering algorithm, please see lower_double_buffer.h. 

Interesting tests are:

- `FusionDoubleBuffering9`: Just adds `1` to a tensor but with nested double buffering at both gmem to smem and smem to registrer
- `FusionSmemBlockGemmCacheDoubleBuffer`: Based on `FusionSmemBlockGemmCache`, but double-buffers both `A` and `B` at both gmem to smem and smem to register.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 1]
```diff
@@ -120,7 +120,8 @@ TensorView::TensorView(const TensorView* src, IrCloner* ir_cloner)
       compute_at_pos_(src->compute_at_pos_),
       max_producer_pos_(src->max_producer_pos_),
       memory_type_(src->memory_type_),
-      swizzle_type_(src->swizzle_type_) {
+      swizzle_type_(src->swizzle_type_),
+      is_double_buffered_(src->is_double_buffered_) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:129]: This will have to be added to the mutator

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 2]
```diff
@@ -891,6 +893,14 @@ void TensorView::clearReductionIterDomains() {
   setDomain(IrBuilder::create<TensorDomain>(container(), new_root, new_contig));
 }
 
+void TensorView::doubleBuffer() {
+  auto def = dynamic_cast<UnaryOp*>(definition());
+  TORCH_CHECK(
+      def != nullptr && def->getUnaryOpType() == UnaryOpType::Set,
+      "Double buffering is only supported when a TensorView is defined in a fusuion with the Set unary op.");
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Is it supported for global memory? Seems that wouldn't make sense?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 3]
```diff
@@ -2096,8 +2139,10 @@ std::vector<Val*> Index::getProducerStridedIndices(
         getNonGlobalProducerStridedIndices(producer, consumer, loops);
   }
 
-  TORCH_INTERNAL_ASSERT(
-      strided_indices.size() == producer->getMaybeRFactorDomain().size());
+  if (!producer->isDoubleBuffered()) {
+    TORCH_INTERNAL_ASSERT(
+        strided_indices.size() == producer->getMaybeRFactorDomain().size());
+  }
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: is this `else {...strided_indices.size() == producer->getMaybeRFactorDomain().size()+1...}`?

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,529 @@
+#include <torch/csrc/jit/codegen/cuda/lower_double_buffer.h>
+
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp]: `toString` doesn't require `ir_iostream.h`, we can generally stop including this if we consistently use `toString`

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:3]: Nit:
I've been trying to be consistent of grouping headers in the order of:
anything inside `torch/csrc/jit/codegen/cuda/` except the header of this specific file
the header for this file
anything inside aten or torch but outside codegen/cuda
standard headers

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+    // Like vectorization, only UnaryOp::Set with another TensorView is
+    // considered.
+    auto def = tv->definition();
+    TORCH_INTERNAL_ASSERT(
+        def->isA<UnaryOp>(),
+        "Invalid tensor to double-buffer. Only tensor defined by UnaryOp::Set is supported: ",
+        def->toString());
+
+    TORCH_INTERNAL_ASSERT(
+        def->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Set,
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp]: Nit: I think this assert can be merged with the one above since message is the same.

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+  return std::any_of(exprs.begin(), exprs.end(), [](const UnaryOp* uop) {
+    return uop->in()->as<TensorView>()->getMemoryType() == MemoryType::Shared;
+  });
+}
+
+// Replicates double buffer loops for Prologue, Main, and
+// Epilogue. Prologue only copies the load expressions of double
+// buffered tensors, whereas Epilogue does any expression other than
+// the loads. Main copies everything.
+class DoubleBufferLoopCloner : public kir::IrVisitor {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:154]: We should probably rename `kir::IrVisitor` to `ExprListVisitor` or something else in a follow up PR to distinguish it from "regular" `IrVisitor`

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:154]: Maybe `KernelIrVisitor`

**Review by csarofeen — APPROVED** [review]
LGTM minor comments. I scanned the tests but expect @shmsong to approve/confirm the code generated for them.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: I saw the checks in lower_shift on memory types. I think it would be good to have them here as well so we can error during scheduling if anything is wrong.

File: test/cpp/jit/test_gpu.cpp [hunk 8]
```diff
@@ -19688,6 +19688,456 @@ TEST_F(NVFuserTest, FusionIssue1305Repro_CUDA) {
...
+  tv2->doubleBuffer();
+
+  tv3->split(-1, 128);
+  tv3->split(-1, 32);
+  TransformPropagator::from(tv3);
+
+  tv0->computeAt(tv3, 1);
+
+  // tv2 is invalid to double-buffer as its producer, tv1, is
+  // computed inside the double-buffering loop.
```
Comment (shmsong) [inline test/cpp/jit/test_gpu.cpp:19874]: Overlooked this constraint in the first pass:

Do we need to require ca pos of tv1 to be < tv pos of tv2? Is it because we always double buffer across the inner most loop? A counter example would be very helpful here. 

It'd also be very helpful if we could support at least ca pos of tv1==ca pos of tv2.

File: test/cpp/jit/test_gpu.cpp [hunk 9]
```diff
@@ -19688,6 +19688,456 @@ TEST_F(NVFuserTest, FusionIssue1305Repro_CUDA) {
...
+  auto t0 = at::randn({200}, options);
+  auto cg_outputs = fe.runFusion({t0});
+
+  auto ref = t0 + 1;
+
+  testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
+}
+
+// Multiple tensors to double-buffer
+TEST_F(NVFuserTest, FusionDoubleBuffering8_CUDA) {
```
Comment (shmsong) [inline test/cpp/jit/test_gpu.cpp:20061]: Do we support double buffering 2 tensors that are a producer-consumer pair across the same loop? Or more generally any dependency in the same double buffer loop.

Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:20061]: This case would make the prolog and epilog a bit different and the switch variable might need to be flipped. Might want to check against this case if not supported.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 10]
```diff
@@ -1297,9 +1303,13 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
     }
   }
 
+  // TODO-DB
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: nitpick: comment

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 11]
```diff
@@ -1510,15 +1520,20 @@ std::vector<Val*> Index::getNonGlobalProducerStridedIndices(
     }
   }
 
+  // TODO-DB: dup
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: nitpick: comment.

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+  using kir::IrVisitor::handle;
+
+  void clone() {
+    const auto gpu_lower = GpuLower::current();
+    kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+    // Cloning the double buffer loop as follows:
+    //
+    // Prologue: 0 to 1
+    // Main: 0 to (extent-1)
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:183]: nit: `Main: 1 to (extent-1)` ?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: I didn't check the memory type here so that `TensorView::setMemoryType()` could be done after `TensorView::doubleBuffer()`. I thought that's more consistent with, say, parallelization where shared memory needs to be used, but we don't check the memory type when setting parallel types. The only condition that won't change at this point is how the tensor is defined, so that's why it's the only one checked here.

I don't have a strong opinion here. Let me know if the memory type should be checked here.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:129]: Sorry, could you be more specific?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: Oops, sorry, this is a temporary change. I'll change the same assertion in the other indexing functions as well.

**Review by shmsong — COMMENTED** [review]
overall looks good to me. Had some questions on a few boundary cases.

File: test/cpp/jit/test_gpu.cpp (hunk 8 above)
Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp:19874]: In my understanding of double buffering, it is required. To be more precise, to double-buffer the load from `tv1` to `tv2`, `tv1` must be defined for the iteration domain where `tv2` is double buffered. That's why the CA position of `tv1` must be left of that of `tv2`. The double buffer axis of `tv2` can be further left if `axis(ca_pos - 1)` is parallelized or unrolled, and in that case the CA position of `tv1` must be further left.

What would the code look like with ca pos of tv1==ca pos of tv2?

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp (hunk 12 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:183]: The main loop spans from 0 to `extent - 1`. The prologue loop just loads the first value but does not do any other computations in the double buffer loop.

File: test/cpp/jit/test_gpu.cpp (hunk 9 above)
Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp:20061]: In `FusionDoubleBuffering9`, `tv2` and `tv3` are double buffered, and `tv2` is the producer of `tv3`. Their double buffer loops are not the same. `tv3`'s double buffering is done inside `tv2`'s CA position. 

If two double-buffered tensors have a producer-consumer dependency, their double buffer axes cannot be the same due to the requirement of the CA positions as discussed in the above comment. So, no such double-buffered tensors can have the same double buffer axis. `DoubleBufferFusionInspector::validate(TensorView*)` should detect such tensors if exist.

File: torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp (hunk 12 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp:183]: I see. Sorry was confused about load iter and op iter. Thanks for clarification.

File: test/cpp/jit/test_gpu.cpp (hunk 9 above)
Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:20061]: I see. It should work fine if consumer can be assumed always on the right of producer.

 This is now the same question as the above one about CA axis constraint. 

Thanks for clarification.

File: test/cpp/jit/test_gpu.cpp (hunk 8 above)
Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:19874]: This seems to be from definition of double buffered iter domains, and I guess the ca pos could be moved by schedulers to fit in the requirement. 

Just wanted to understand more about scheduling under this requirement, mainly wanted to confirm if we could generate the below two cases:

My understanding for the `==` case is like: 
```cpp
  // T1[i, i(8)], T2[i, i(8)], both ca pos 1, prolog not included
float t2[8*2]; // double buffered
for idx0 in ...
  float t1[8]
  for idx1 in 0..8:
    t1[idx1]=...
  for idx2 in 0..8:
    t2[idx2+8*(1-idx0%2)] = t1[idx2]

  ... (use t2[idx2+8*idx0%2])
```
This one could be seen very often and would be great if we could support.


ca pos of tv1 > ca pos of tv2 case could be 
```cpp
 // T1[i, i(8)] (ca -1), T2[i, i(8)] (ca 1) prolog not included
float t2[8*2]; // double buffered
for idx0 in ...
  for idx2 in 0..8:
   float t1[0];
   t1[0] = ... + ...
   t2[idx2+8*(1-idx0%2)] = t1[0]

  ... (use t2[idx2+8*idx0%2])
```
This might be useful in prolog fusion but seems low priority for now.

Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp:19874]: In the first example, when assigning to `t2`, the logical indices of `t2` are `[idx0+1, idx2]`, and when `t2` is used, they are `[idx0, idx2]`, right? This means that the RHS of the assignment, `t1`, must also point to `[idx0+1, idx2]`, but `t1` is defined for `[idx0, idx2]`, so it doesn't work.

I did think about recursively changing the index of the direct and indirect producers of a double buffer tensor, but that seems like a challenge, and I'm not sure it's something we would really need.

In the first case, instead of double buffering `t2`, why not `t1`? If `t1` is defined with an input global memory tensor, double buffering `t1` should be fine.

Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:19874]: I see the challenge now. Thanks for pointing this out. Would it be possible to support maybe just a limited set of cases? maybe just a chain of single-use unary ops all the way back to gmem input would be all I need for now. 

Doesn't have to be on the same PR though.

Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp:19874]: Can you show an actual example code? If it's just a chain of unary ops leading to gmem, why not just double buffer the first load from the gmem?

Comment (shmsong) [reply test/cpp/jit/test_gpu.cpp:19874]: Tracking in #1397.

**Review by shmsong — APPROVED** [review]
LGTM. 

This is very useful and powerful perf feature. Thanks for building it and enabling this optimization! 👍
